src/sources/docker.rs

```diff
@@ -379,256 +462,121 @@ impl DockerSource {
-        let mut state = Some((self.main_send.clone(), info));
+        let mut info = Some(info);
+        let main_send = self.main_send.clone();
         let partial_event_marker_field = self.core.config.partial_event_marker_field.clone();
         let auto_partial_merge = self.core.config.auto_partial_merge;
         let mut partial_event_merge_state = None;
-        tokio01::prelude::stream::poll_fn(move || {
-            // !Hot code: from here
-            if let Some(&mut (_, ref mut info)) = state.as_mut() {
-                // Main event loop
-                loop {
-                    return match stream.poll() {
-                        Ok(Async::Ready(Some(message))) => {
-                            if let Some(event) = info.new_event(
-                                message,
-                                partial_event_marker_field.clone(),
-                                auto_partial_merge,
-                                &mut partial_event_merge_state,
-                            ) {
-                                Ok(Async::Ready(Some(event)))
-                            } else {
-                                continue;
-                            }
-                            // !Hot code: to here
-                        }
-                        Ok(Async::Ready(None)) => break,
-                        Ok(Async::NotReady) => Ok(Async::NotReady),
-                        Err(error) => {
-                            match error {
-                                Error::Fault { code, .. } if code == StatusCode::NOT_IMPLEMENTED => {
-                                    error!(r#"docker engine is not using either `jsonfile` or `journald`
-                                            logging driver. Please enable one of these logging drivers
-                                            to get logs from the docker daemon."#);
-                                    break;
-                                }
-                                error => {
-                                    error!(message = "docker API container logging error",%error);
-                                    // On any error, restart connection
-                                    break;
-                                }
+
+        stream
+            .map(|value| {
+                match value {
+                    Ok(message) => Ok(info.as_mut().unwrap().new_event(
+                        message,
+                        partial_event_marker_field.clone(),
+                        auto_partial_merge,
+                        &mut partial_event_merge_state,
+                    )),
+                    Err(error) => {
+                        // On any error, restart connection
+                        match error.kind() {
+                            // http::StatusCode::NOT_IMPLEMENTED
+                            DockerErrorKind::DockerResponseServerError { status_code, .. }
+                                if *status_code == 501 =>
```

Comment:
Can we use `StatusCode::NOT_IMPLEMENTED as u32` or something like that so to avoid the need for the above comment.

Reply:
Changed to constant from `http` crate :+1: